pkg/aws_keyhub/keyhub.go

```diff
@@ -108,37 +124,197 @@ func PollForAccessToken(authorizeDeviceresponse AuthorizeDeviceResponse, noOfTim
+		if err != nil {
+			logrus.Errorln("Failed to read KeyHub response body:", err)
+		}
+		defer resp.Body.Close()
+		logrus.Fatal("Received HTTP response body:", string(body))
 	}
 
-	var result LoginResponse
-	if err := json.Unmarshal(body, &result); err != nil { // Parse []byte to the go struct pointer
+	var result AuthResponse
+	if err := json.Unmarshal(body, &result); err != nil {
 		logrus.Fatal("Error, cannot unmarshal JSON")
 	}
 
+	if result.RefreshToken != nil {
+		storeRefreshToken(result)
+	}
+
 	return result
 }
 
-func ExchangeToken(loginResponse LoginResponse) ExchangeResponse {
+func storeRefreshToken(authResponse AuthResponse) {
+	logrus.Debugln("Storing refresh token to file.")
+	filePath := GetAwsKeyHubRefreshTokenPath()
+	writeFile, err := os.OpenFile(filePath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0600)
+
+	refreshTokenFile := RefreshTokenFile{
+		RefreshToken: *authResponse.RefreshToken,
+		ExpireDate:   time.Now().Add(time.Duration(authResponse.ExpiresIn) * time.Second),
+	}
+	if err != nil {
+		logrus.Fatal("Error creating refresh-token.json file:", err)
+	}
+	defer writeFile.Close()
+	jsonBytes, err := json.Marshal(refreshTokenFile)
+	if err != nil {
+		logrus.Fatal("Error marshaling refresh token to JSON:", err)
+	}
+	_, err = writeFile.Write(jsonBytes)
+	if err != nil {
+		logrus.Fatal("Error writing to refresh-token.json file:", err)
+	}
+	logrus.Debugln("Wrote refresh token to file at", filePath)
+}
+
+func DoLogin() AuthResponse {
+	// Check if access token exists and is still valid
+	refreshTokenFile := readRefreshToken()
+	var authResponse *AuthResponse
+	if refreshTokenFile != nil && isAccessTokenValid(*refreshTokenFile) {
+		logrus.Infoln("Attempting KeyHub login using refresh token.")
+		authResponse = getAccessTokenWithRefreshToken(*refreshTokenFile)
+	} else if refreshTokenFile != nil {
+		removeInvalidOrExpiredRefreshTokenFile()
+	}
+
+	logrus.Debugln("AuthResponse from refresh token:", authResponse)
+
+	if authResponse != nil {
+		return *authResponse
+	}
+	authorizeDeviceResponse := authorizeDevice()
+	return pollForAccessToken(authorizeDeviceResponse, 0)
+
+}
+
+func readRefreshToken() *RefreshTokenFile {
+	filePath := GetAwsKeyHubRefreshTokenPath()
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil
+	}
+	defer file.Close()
+
+	var refreshTokenFile RefreshTokenFile
+	decoder := json.NewDecoder(file)
+	if err := decoder.Decode(&refreshTokenFile); err != nil {
+		logrus.Fatal("Error decoding refresh-token.json file:", err)
+	}
+	return &refreshTokenFile
+}
+
+func isAccessTokenValid(refreshTokenFile RefreshTokenFile) bool {
+	if refreshTokenFile.RefreshToken == "" {
+		return false
+		// If the token is expired or will expire in the next 30 seconds, consider it invalid
+	} else if refreshTokenFile.ExpireDate.Add(-30 * time.Second).Before(time.Now()) {
+		return false
+	}
+
+	return true
+}
+
+func getAccessTokenWithRefreshToken(refreshTokenFile RefreshTokenFile) *AuthResponse {
```

Comment:
This function contains a lot of logic. Maybe we can move the handling of the responsecode 400 and default out of it? Also preparing the payload and posting could be separated.